Proposals/NNNN-concurrency-safe-notifications.md

```diff
@@ -0,0 +1,386 @@
+
+Existing code which vends notifications do not need to alter existing `Notification` declarations, observers, or posts to adopt to this proposal.
+
+```swift
+extension NSWorkspace {
+    public struct WillLaunchApplication: NotificationCenter.MainActorMessage {
+        public static var name: Notification.Name { NSWorkspace.willLaunchApplicationNotification }
+        public typealias Subject = NSWorkspace
+        
+        public var application: NSRunningApplication
+
+        public static func makeMessage(_ notification: Notification) -> Self? {
+            guard
+                let application = notification.userInfo?["applicationUserInfoKey"] as? NSRunningApplication
+            else {
+                return nil
+            }
+            
+            return Self(application: application)
+        }
+        
+        public static func makeNotification(_ message: Self) -> Notification {
+            return Notification(name: Self.name, userInfo: ["applicationUserInfoKey": message.application])
+        }
+    }
+}
+
+extension NotificationCenter.MessageIdentifier where Self == NotificationCenter.BaseMessageIdentifier<NSWorkspace.WillLaunchApplication> {
+    static var willLaunchApplication: Self { .init() }
+}
+```
+
+This notification could be observed by a client using:
+
+```swift
+let token = center.addObserver(of: workspace, for: .willLaunchApplication) { message in
+    // Do something with message.application
+}
+
+// Or, without a specific instance to observe ...
+
+let token = center.addObserver(of: NSWorkspace.self, for: .willLaunchApplication) { message in
+    // Do something with message.application
+}
+```
+
+And it could be posted using:
+
+```swift
+NotificationCenter.default.post(
+    NSWorkspace.WillLaunchApplication(application: launchedApplication),
+    with: workspace
+)
+```
+
+## Detailed design
+
+### `NotificationCenter.Message`, `NotificationCenter.MainActorMessage`, and `NotificationCenter.AsyncMessage`
+
+The `NotificationCenter.Message` protocol acts as a base for `NotificationCenter.MainActorMessage` and `NotificationCenter.AsyncMessage`, helping the two share functionality:
+
+```swift
+@available(FoundationPreview 0.5, *)
+extension NotificationCenter {
+    public protocol Message {
+        associatedtype Subject: AnyObject
+        static var name: Notification.Name { get }
+        
+        static func makeMessage(_ notification: Notification) -> Self?
+        static func makeNotification(_ message: Self) -> Notification
+    }
+    
+    public protocol MainActorMessage: Message {}
```

Comment:
I am wondering if this protocol should have a `@MainActor` annotation. Those types will only be sent and received from the MainActor anyways.